File: 백준/Gold/16236. 아기 상어/아기 상어.py
import sys
from collections import deque

N = int(input())
board = []
shark_size = 2
shark_stk = 0
sx, sy = 0, 0

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
for i in range(N):
    _li = list(map(int, input().split()))
    for j in range(N):
        if _li[j] == 9:
            _li[j] = 0
            sx, sy = i, j
    board.append(_li)


def isValid(nx, ny):
    return 0 <= nx < N and 0 <= ny < N



ans = 0
while True:
    visited = [[-1 for _ in range(N)] for _ in range(N)]
    q = deque()
    q.append([sx, sy])
    visited[sx][sy] = 0
    cand = list()
    flag = False
    while q:
        x, y = q.popleft()

        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if isValid(nx, ny) and visited[nx][ny] == -1:
                # 자신보다 큰 물고기
                if board[nx][ny] > shark_size:
                    continue
                
                # 자신이랑 같은 물고기
                elif board[nx][ny] == shark_size or board[nx][ny] == 0:
                    q.append((nx, ny))
                    visited[nx][ny] = visited[x][y] + 1
                # 먹을 수 있는 물고기
                else:
                    flag = True
                    visited[nx][ny] = visited[x][y] + 1
                    cand.append((visited[nx][ny], nx, ny))
    if not flag:
        break
    else:
        cand.sort()
        ans += cand[0][0]
        sx, sy = cand[0][1], cand[0][2]
        shark_stk += 1
        if shark_stk == shark_size:
            shark_size += 1
            shark_stk = 0
        board[sx][sy] = 0

    
print(ans)

File: 백준/Gold/16236. 아기 상어/아기 상어.cc
#include <bits/stdc++.h>
#define endl "\n"
#define ll long long
#define INF 987654321
#define MAX 21
#define MOD 1000000000
#define PI 3.1415926254
using namespace std;
typedef pair<int, int> p;

int n, shark_size = 2, ans = 0;
int cnt = 0;
int visited[MAX][MAX];
int board[MAX][MAX];
vector<pair<int, pair<int, int>>> ord;	//거리 정보, 좌표
p shark_coord;

p Dir[4] = { {-1, 0}, {0, -1}, {1, 0}, {0, 1} };


bool cmp(pair<int, pair<int, int>> a, pair<int, pair<int, int>> b) {
	if (a.first == b.first) {
		if (a.second.first == b.second.first)
			return a.second.second < b.second.second;
		return a.second.first < b.second.first;
	}
	return a.first < b.first;
}
int main() {
	ios::sync_with_stdio(false);
	cin.tie(0); cout.tie(0);

	cin >> n;
	for(int i = 0; i <n; ++i)
		for (int j = 0; j < n; ++j) {
			cin >> board[i][j];
			if (board[i][j] == 9) {
				board[i][j] = 0;
				shark_coord.first = i;
				shark_coord.second = j;
			}
		}
	while (1) {

		memset(visited, -1, sizeof(visited));
		bool found = false;

		queue<p> q;
		q.push({shark_coord.first, shark_coord.second});
		visited[shark_coord.first][shark_coord.second] = 0;

		while (!q.empty()) {

			int x = q.front().first; int y = q.front().second;
			q.pop();

			for (int i = 0; i < 4; ++i) {
				int nx = x + Dir[i].first; int ny = y + Dir[i].second;
				if (0 <= nx && nx < n && 0 <= ny && ny < n && visited[nx][ny] == -1) {

					//더 큰 상어
					if (board[nx][ny] > shark_size) continue;

					//상어 발견
					if (board[nx][ny]  < shark_size && board[nx][ny] != 0) {
						found = true;
						visited[nx][ny] = visited[x][y] + 1;
						ord.push_back({ visited[nx][ny], {nx, ny} });
						continue;
					}

					//그냥 지나갈 경우
					visited[nx][ny] = visited[x][y] + 1;
					q.push({ nx, ny });

				}
			}
		}
		if (!found) break;
		else {
			sort(ord.begin(), ord.end(), cmp);
			ans += ord[0].first;
			shark_coord = ord[0].second;
			cnt++;
			if (cnt == shark_size) {
				shark_size++; cnt = 0;
			}
			board[shark_coord.first][shark_coord.second] = 0;
			ord.clear();
		}
	}
	cout << ans << endl;
	return 0;
}